superset/alfred/utils.py

```diff
@@ -0,0 +1,350 @@
+from alfred_client import AlfredService
+from alfred_client.instance import InstanceUtil
+from alfred_client import RetentionInstance
+
+import logging
+import os.path
+from datetime import datetime
+from datetime import timedelta
+from typing import Optional, Union
+import json
+
+from alfred_client.instance import InstanceUtil
+from alfred_client.instance.AbstractInstance import AbstractInstance
+
+# Determine if we are in JupyHub Hogwarts environment by the presence of the hogwarts library.
+
+import jwt
+
+logger = logging.getLogger(__name__)
+
+# limit to 100 emls retained at a time
+limit = 100
+
+def validate_json(value: Union[bytes, bytearray, str]) -> None:
+    try:
+        utils.validate_json(value)
+    except SupersetException as ex:
+        raise ValidationError("JSON not valid") from ex
+
+class TokenOAuthInstance(AbstractInstance):
+    """
+    Alfred Instance with OAuth with on-behalf-of authentication instance only supported from Hogwarts. This
+    implementation will fail if the hogwarts library is not present. This is a fair indication that the user is not in
+    the JH environment.
+    """
+
+    def __init__(self, url: str, token: str):
+        """
+        Constructor.
+
+        Args:
+            url:
+                The url to alfred service.
+        """
+        super(TokenOAuthInstance, self).__init__(url)
+        self.__access_token: str = token
+        self.__oath_token_expiry: Optional[datetime] = None
+        self.__oath_token: Optional[str] = None
+
+    def _add_header(self):
+        expiry_buffer = timedelta(minutes=30)
+
+        if (
+            self.__oath_token_expiry is None
+            or self.__oath_token_expiry - datetime.now() < expiry_buffer
+        ):
+            logger.info("Authenticating via hogwarts vault...")
+            api = InstanceUtil.build_scope(self.url)
+            vault_client = VaultClient()
+            # jh_ctx_refresh_token = ContextualRefreshToken(os.environ.get('VAULT_REFRESH_TOKEN'))
+
+            # vault_client.refresh(jh_ctx_refresh_token)
+
+            alfred_app = vault_client.get_oauth_client("alfred")
+            obo_access_token, obo_ctx_refresh_token = vault_client.on_behalf_of(
+                f"{alfred_app.audience}/{alfred_app.scope}",
+                self.__access_token,
+                token_client_name="superset",
+            )
+
+            refreshed_obo_access_token = vault_client.refresh(obo_ctx_refresh_token)
+            self.__oath_token = refreshed_obo_access_token[0]
+            decoded = jwt.decode(
+                str.encode(self.__oath_token),
+                options={"verify_signature": False, "verify_aud": False},
+            )
+            self.__oath_token_expiry = datetime.fromtimestamp(decoded["exp"])
+
+            logger.info("Authenticated.")
+
+        self.session.headers["Authorization"] = f"Bearer {self.__oath_token}"
+
+def create_retention(
+    data_set,
+    alfred_instance,
+    metadata_system="HOGWARTS",
```

Comment:
```suggestion
    metadata_system="Analytical Platform",
```

Maybe? Or would "Superset" be appropriate?